efficacy/app.R

```diff
@@ -11,52 +11,65 @@ options(shiny.useragg = FALSE)
 
 nest_logo <- "https://raw.githubusercontent.com/insightsengineering/hex-stickers/main/PNG/nest.png"
 
-# code>
-## Generate Data
-ADSL <- synthetic_cdisc_data("latest")$adsl
-adsl_labels <- teal.data::col_labels(ADSL, fill = FALSE)
+## Data reproducible code ----
+data <- teal_data()
+data <- within(data, {
+  ADSL <- synthetic_cdisc_data("latest")$adsl
+  adsl_labels <- teal.data::col_labels(ADSL, fill = FALSE)
 
-char_vars_asl <- names(Filter(isTRUE, sapply(ADSL, is.character)))
+  char_vars_asl <- names(Filter(isTRUE, sapply(ADSL, is.character)))
 
-adsl_labels <- c(
-  adsl_labels,
-  AGEGR1 = "Age Group"
-)
-ADSL <- ADSL %>%
-  mutate(
-    AGEGR1 = factor(case_when(
-      AGE < 45 ~ "<45",
-      AGE >= 45 ~ ">=45"
-    ))
-  ) %>%
-  mutate_at(char_vars_asl, factor)
+  adsl_labels <- c(
+    adsl_labels,
+    AGEGR1 = "Age Group"
+  )
+  ADSL <- ADSL %>%
+    mutate(
+      AGEGR1 = factor(case_when(
+        AGE < 45 ~ "<45",
+        AGE >= 45 ~ ">=45"
+      ))
+    ) %>%
+    mutate_at(char_vars_asl, factor)
 
-teal.data::col_labels(ADSL) <- adsl_labels
+  teal.data::col_labels(ADSL) <- adsl_labels
 
-ADTTE <- synthetic_cdisc_data("latest")$adtte
+  ADTTE <- synthetic_cdisc_data("latest")$adtte
 
-ADRS <- synthetic_cdisc_data("latest")$adrs
-adrs_labels <- teal.data::col_labels(ADRS, fill = FALSE)
-ADRS <- filter(ADRS, PARAMCD == "BESRSPI" | AVISIT == "FOLLOW UP")
-teal.data::col_labels(ADRS) <- adrs_labels
+  ADRS <- synthetic_cdisc_data("latest")$adrs
+  adrs_labels <- teal.data::col_labels(ADRS, fill = FALSE)
+  ADRS <- filter(ADRS, PARAMCD == "BESRSPI" | AVISIT == "FOLLOW UP")
+  teal.data::col_labels(ADRS) <- adrs_labels
 
-ADQS <- synthetic_cdisc_data("latest")$adqs
-adqs_labels <- teal.data::col_labels(ADQS, fill = FALSE)
-ADQS <- ADQS %>%
-  filter(ABLFL != "Y" & ABLFL2 != "Y") %>%
-  filter(AVISIT %in% c("WEEK 1 DAY 8", "WEEK 2 DAY 15", "WEEK 3 DAY 22")) %>%
-  mutate(
-    AVISIT = as.factor(AVISIT),
-    AVISITN = rank(AVISITN) %>%
-      as.factor() %>%
-      as.numeric() %>%
-      as.factor()
-  )
-teal.data::col_labels(ADQS) <- adqs_labels
+  ADQS <- synthetic_cdisc_data("latest")$adqs
+  adqs_labels <- teal.data::col_labels(ADQS, fill = FALSE)
+  ADQS <- ADQS %>%
+    filter(ABLFL != "Y" & ABLFL2 != "Y") %>%
+    filter(AVISIT %in% c("WEEK 1 DAY 8", "WEEK 2 DAY 15", "WEEK 3 DAY 22")) %>%
+    mutate(
+      AVISIT = as.factor(AVISIT),
+      AVISITN = rank(AVISITN) %>%
+        as.factor() %>%
+        as.numeric() %>%
+        as.factor()
+    )
+  teal.data::col_labels(ADQS) <- adqs_labels
+})
+
+# set datanames
+datanames <- c("ADSL", "ADTTE", "ADRS", "ADQS")
+datanames(data) <- datanames
 
-# <code
+# set join_keys
+data@join_keys <- cdisc_join_keys(!!!datanames)
```

Comment:
we need to fix this:

We need to reconsider following functions to have a robust and easy api.
- `join_keys`
- `cdisc_join_keys`
- `get_cdisc_keys`
- `get_join_keys`


First though is:
```r
jk <- join_keys() # constructor
join_keys(jk) <- # setter
join_keys(jk) # getter
```

setting `join_keys` should be possible through:
```
join_keys(data)["dataset", "dataset"] <- character
join_keys(data) <- join_keys # set or replace whole keys object
```

cdisc_join_keys is just a wrapper around join_keys constructor - not possible to set nor get keys using this function. Most probably should have the same assertion for `...` as `join_keys`

```
cdisc_join_keys()
```

`get_cdisc_keys()` could change - now it returns a list, but it could return JoinKeys. Unfortunately, `get_join_keys()` is used widely across all packages. We will need to make a relevant changes there also (teal.transform :()
```r
get_join_keys(c("ADSL", "ADTTE", ...)) # returns JoinKeys
```

Reply:
How about getting and setting keys on `cdisc` data with `join_keys` methods?

Reply:
> How about getting and setting keys on cdisc data with join_keys methods?

Not sure what you mean. 

----

The way I see setting of the `JoinKeys`

```
data <- teal_data()
data <- within(data, {
  ADSL <- synthetic_cdisc_data("latest")$adsl
  ADRS <- synthetic_cdisc_data("latest")$adrs
  ADLB <- synthetic_cdisc_data("latest")$adlb
  ADELO <- ADLB
})
join_keys(data) <- join_keys() # technically it's (and should be) a default in teal_data()

datanames <- c("ADSL", "ADRS", "ADLB", "ADELO")
join_keys(data) <- default_cdisc_keys(datanames) # should warn about missing keys for any of the datasets
join_keys(data)["ADELO"] <- get_cdisc_keys("ADLB")
join_keys(data)["ADELO", "ADSL"] <- c("STUDYID", "USUBJID")

```

Reply:
If `cdisc_data` is dropped as a class, this is almost exactly what I meant.
Can we have one function to generate the keys? `join_keys(..., cdisc = FALSE)` I may be missing something.

Reply:
I don't know. I'm thinking there should be `cdisc_data` package with `cdisc_join_keys` and `cdisc_data` function maybe?

Reply:
WHAT IF `JoinKeys` will be just a list with S3 class?? 
We could then drop many dependencies in other packages where JoinKeys are needed (teal.transform)

Reply:
Sounds nice :+1: